### PR TITLE
raft: plumbing crdb version handle into raft

### PR DESCRIPTION
**Pull request by lyang24**
We need the ability to safely ship backward incompatible changes in the raft module. This commit plumbs version handle in to raft package to enable version gate infrastructure in raft.

Epic: None

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/store.go [hunk 1]
```diff
@@ -360,6 +360,7 @@ func newRaftConfig(
 	appliedIndex kvpb.RaftIndex,
 	storeCfg StoreConfig,
 	logger raft.Logger,
+	crdbVersion *clusterversion.Handle,
```
Comment (pav-kv) [inline pkg/kv/kvserver/store.go:363]: This can be taken from `storeCfg`

File: pkg/raft/raft.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ import (
 	"crypto/rand"
 	"errors"
 	"fmt"
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
```
Comment (pav-kv) [inline pkg/raft/raft.go:25]: nit: place this import into the crdb group below

**Comment by pav-kv** [comment]
Probably need to run `dev generate bazel` to update the BUILD files.

**Comment by pav-kv** [comment]
LGTM, this will do. Let's keep this uncommitted for now, it's better to include this commit into a PR that would use the gate.

**Review by lyang24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122165)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/raft_test.go` line 1149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122165#-NvIWGEY39L5LtTVMgX8:-NvIWGEY39L5LtTVMgX9:b8fahe6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb863dee94fdd33d086b8b3465301f5cada2a924/pkg/raft/raft_test.go#L1149)):*
> ```Go
> 
> 		// Increase the commit index only if the log is in sync with the leader.
> 		{pb.Message{From: 2, To: 1, Type: pb.MsgHeartbeat, Term: 2, Commit: commit + 1}, false, commit},
> ```

I copied code from this commit https://github.com/etcd-io/raft/pull/140/commits/eb6bfc6966a2e90589986c8f6fd808d6d97ab1e9 
This test case will fail I see both leader's commit index and follower's last index is at 3 and we are expecting 2 
cc @pav-kv

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122165)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`pkg/raft/raft_test.go` line 1149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122165#-NvIWGEY39L5LtTVMgX8:-NvbR-Tw8C38HpxO3BPE:b-pztnod) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb863dee94fdd33d086b8b3465301f5cada2a924/pkg/raft/raft_test.go#L1149)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

I copied code from this commit https://github.com/etcd-io/raft/pull/140/commits/eb6bfc6966a2e90589986c8f6fd808d6d97ab1e9 
This test case will fail I see both leader's commit index and follower's last index is at 3 and we are expecting 2 
cc @pav-kv
</blockquote></details>

This test was buggy initially, let's clean it up first (see the other comment).
___
*[`pkg/raft/raft_test.go` line 1158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/122165#-NvbPHgVEAK9GzI7Srxu:-NvbPHgVEAK9GzI7Srxv:be1c20z) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb863dee94fdd33d086b8b3465301f5cada2a924/pkg/raft/raft_test.go#L1158)):*
> ```Go
> 		require.NoError(t, storage.Append(index(1).terms(1, 2, 3)))
> 		sm := newTestRaft(1, 5, 1, storage)
> 		sm.becomeFollower(2, 2)
> ```

This test was incorrect. The `sm.becomeFollower(2, 2)` call makes the node a follower at term 2 with the leader node ID = 2. But the log in the `Storage`, initialized 2 lines above, ends with an entry at term 3. This can't be the case: the invariant is that `Term >= log.Entry(lastIndex).Term`.

We should probably:
- Call `storage.SetHardState()` with the `HardState` at term 3, and commit index at `commit`.
- Then the `newTestRaft`, while creating the raft node, will read the hard state from storage and init it correctly.
- Then we can remove the `commitTo` invocation in the next line.
- And we probably can remove the `logSynced` initialization as well, because it should be initialized when the raft node is created in `newTestRaft`.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122165)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`pkg/raft/raft.go` line 358 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/122165#-NvlLMAL4SGhDhDRPsYD:-NvlLMAL4SGhDhDRPsYE:b-rxy38g) ([raw file](https://github.com/cockroachdb/cockroach/blob/24cdeddf10097a204881547e3d321a64f1374464/pkg/raft/raft.go#L358)):*
> ```Go
> 	// candidate. For a follower, becomes true the first time a MsgApp append to
> 	// the log succeeds.
> 	logSynced bool
> ```

Let's replace this field with `accTerm uint64` (the latest "accepted term").

The code for tracking this field will be simpler because `accTerm` will be update only in 2 places: when accepting a log append, or when installing a snapshot (which wipes the log, and is equivalent to simultaneously accepting and committing the state represented by this snapshot).

In contrast, the code for updating `logSynced` is a bit less explicit, and there are more conditions under which it has to be updated (for example, we need to update it also when `r.Term` is bumped).

Having the `accTerm` field will be aligned with https://github.com/cockroachdb/cockroach/issues/122446, and eventually this field will become persistent.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122165)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`pkg/raft/raft.go` line 358 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/122165#-NvlLMAL4SGhDhDRPsYD:-NvlMvQu5QaI16oVgGuB:b-80t5er) ([raw file](https://github.com/cockroachdb/cockroach/blob/24cdeddf10097a204881547e3d321a64f1374464/pkg/raft/raft.go#L358)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Let's replace this field with `accTerm uint64` (the latest "accepted term").

The code for tracking this field will be simpler because `accTerm` will be update only in 2 places: when accepting a log append, or when installing a snapshot (which wipes the log, and is equivalent to simultaneously accepting and committing the state represented by this snapshot).

In contrast, the code for updating `logSynced` is a bit less explicit, and there are more conditions under which it has to be updated (for example, we need to update it also when `r.Term` is bumped).

Having the `accTerm` field will be aligned with https://github.com/cockroachdb/cockroach/issues/122446, and eventually this field will become persistent.
</blockquote></details>

You could draw some inspiration from the earlier version of the upstream PR: https://github.com/etcd-io/raft/commit/87ac09e77c2185ab0adfa1f4d634e058a976c5d0.

**Pull request closed without merging**